src/lib/pubkey/pk_options.h

```diff
@@ -0,0 +1,137 @@
+      ///
+      /// For SM2 this context is the user identifier
+      ///
+      /// If the scheme does not support contextual identifiers, then an exception
+      /// will be thrown.
+      PK_Signature_Options with_context(std::span<const uint8_t> context) const;
+
+      /// Specify a context as a string
+      ///
+      /// Equivalent to the version taking a span above; just uses the bytes
+      /// of the string instead.
+      PK_Signature_Options with_context(std::string_view context) const;
+
+      /// Request producing a deterministic signature
+      ///
+      /// Some signature schemes are always deterministic, or always randomized.
+      /// Others support both randomized or deterministic options. This allows
+      /// requesting this. For signatures which are always deterministic or
+      /// always randomized, this option has no effect.
+      ///
+      /// This option is ignored for verification
+      PK_Signature_Options with_deterministic_signature() const;
+
+      /// Specify producing or expecting a DER encoded signature
+      ///
+      /// This is mostly used with ECDSA
+      ///
+      /// For schemes that do not support such formatting (such as RSA
+      /// or post-quantum schemes), an exception will be thrown when the
+      /// PK_Signer or PK_Verifier is created.
+      PK_Signature_Options with_der_encoded_signature() const;
+
+      /// Specify a provider that should be used
+      ///
+      /// This is rarely relevant
+      PK_Signature_Options with_provider(std::string_view provider) const;
+
+      const std::string& hash_function() const { return m_hash_fn; }
+
+      const std::optional<std::string>& prehash_fn() const { return m_prehash; }
+
+      const std::optional<std::string>& padding() const { return m_padding; }
+
+      const std::optional<std::vector<uint8_t>>& context() const { return m_context; }
+
+      const std::optional<std::string>& provider() const { return m_provider; }
+
+      bool using_der_encoded_signature() const { return m_use_der; }
+
+      bool using_deterministic_signature() const { return m_deterministic_sig; }
+
+      bool using_context() const { return context().has_value(); }
+
+      bool using_prehash() const { return m_use_prehash; }
+
+      bool using_padding() const { return padding().has_value(); }
+
+      bool using_provider() const { return provider().has_value() && provider().value() != "base"; }
+
+      // Returns padding plus hash formatted for RSA
+      std::string _padding_with_hash() const;
+
+      /// This is a compatability interface that parses padding in the context
+      /// of the key type, following internal logic used previously.
+      ///
+      /// This is an internal library function and should not be called by
+      /// applications. It will be removed in Botan4.
+      ///
+      static PK_Signature_Options _parse(const Public_Key& key, std::string_view padding, Signature_Format format);
```

Comment:
There actually is no need for `_parse` to be a member function, I think it just accesses the public interface. [Since if _parse needed to do something that was not possible with the public interface, it would mean there is some configuration of hashes/etc that can be used in the library today that couldn’t be requested using the public API.] I’ll just move this to the internal `impl` header.

Reply:
No longer public/"not public"